Player Abstraction

aberinnj opened the pull request:
# Adding Player Entity

## Tasks
- [ ] Moving meat of Player functions to its own class
- [x] Generalizing leaderboard and making it support other sports

Commit:
Added entities for Player; Added player-specific functionalities

Commit:
Adds 'addGame' method to Player class and creates PlayerTest file

aberinnj commented:
Still needs a bit of work. Please review though if we're going the right direction.

Commit:
Changes Player constructor and methods to 'public' and adds some JUnit tests to PlayerTest

Commit:
Added GameStatsTests; Separated GameStats class; Updated Player functions

namanshenoy commented on src/main/java/Entities/Player.java:
```diff
@@ -0,0 +1,38 @@
+		return null;
+	}
+
+	public void updateGameStatsAfter_WinIn(String sport) {
```
What are the advantages of storing records for a player in the player model vs in the game itself?

aberinnj replied:
You mean have it on all under each games? like so:
"ping pong": [player1, player2, player3],
"soccer": []
?
I was thinking:
```json
"pingpong": {
   "players": ["p1_username", "p2_username", "p3_username"], 
   "teams": [not-yet-explored], 
   "metadata": {"team-limit": 2}, 
   "houses": [not-yet-explored], 
   "brackets":["burning-sun", "lava", "blazing-fire", "smouldering-fire"]
    }
}
```
### Meanwhile in Players
```json
"ready_player_one@oracle.com": {
   "teams": {"pingpong":[], "soccer":[]},
   "house": "TootyFrootie",
   "games": [{
        "pingpong":{"win":1, "loss":24, "winstreak": 1, 
               preferences: "rely-on-honor-system"}
    }],
   "achievement": ["played-all-games", "burning-man"]
}
```
- I guess Players would be for player dashboard and stats, badges, house, personal teams?
- For viewing leaderboards and brackets, we get list of usernames from game records, and query player records for displaying player data? vice versa for the above bullet. 
- We can also store teams in game? and have player details separated, just seems easier to handle/debug; simpler to maintain?
- Separation of concerns I guess. Players keep players and their extras; games keep games and their extras. I don't know why I thought of separating em when I first saw them. Am I thinking like it's relational? 
- Functions handling all these player stuff would stay in Players. {Update win, Join Team, Join House, Change preferences, ManageUserDetailsForTypos}. Functions handling etc game details would stay in Games? {GetPlayerList, DisplayTeams, DisplayTeamPlayers, DisplayHouses, GetSportBrackets, AddNewSport}. Having it separated also has the unnecessary benefit of having smaller classes, or even smaller functions.

*Please do correct me if there's a more effective way or if I am over-engineering it again*

---

Or do you mean something else by ```storing it in the game itself```?

aberinnj replied:
That's not what we wrote though... I see, yeah scope is too small. Blind coding from the beginning. Will fix!

namanshenoy commented on src/main/java/Entities/GameStats.java:
```diff
@@ -0,0 +1,45 @@
+	public int getWinstreak() {
+		return this.winstreak;
+	}
+}
```
New Line is good practice

aberinnj replied:
Is it my git clrf settings? or is there really just no newline after }.

namanshenoy replied:
There's just no newline after }

Commit:
Pat and Norman's work on backend

Among some other changes, Norman and I renamed the "PongService" to
"GameService" so that this leaderboard is more generalizable.  We also
added a "Player" class as a separate file, but we will still need to
move the functionality of the old "Player" class inside of "GameService" to the new, separate
one.  We also began to write some tests in "PlayerTest" and "MainTest,"
but they will need more tests as well.

Commit:
Pat and Norman's work on backend

Among some other changes, Norman and I renamed the "PongService" to
"GameService" so that this leaderboard is more generalizable.  We had
thought about moving the "Player" class from inside the "GameService" file to
its own file so that it wouldn't be so tightly coupled with the former.
This might help in making the leaderboard generalizable to other sports
or activities beyond ping pong.  But we're still trying to figure out the best way to store the name of the activity/sport.
Some players will participate in multiple sports/activities, while others
will likely only participate in one.  So we'll still need to add
functionality for a user of the leaderboard to be able to only show the
leaderboard for a particular sport/activity (such as volleyball). They
should then only see the rankings of players who play volleyball, not
necessarily all the players in the system.
